src/lib/storage/buffer/storage_region.cpp

```diff
@@ -0,0 +1,166 @@
+#include "storage_region.hpp"
+#ifdef __linux__
+#include <linux/fs.h>
+#include <sys/ioctl.h>
+#endif
+
+#include <unistd.h>
+
+namespace hyrise {
+
+inline void DebugAssertPageAlignment(const std::byte* data) {
+  DebugAssert(reinterpret_cast<std::uintptr_t>(data) % 512 == 0,
+              "Data buffer is not properly aligned to 512 byte boundary as required for direct IO: " +
+                  std::to_string(reinterpret_cast<std::uintptr_t>(data) % 512));
+}
+
+static StorageRegion::Mode find_mode_or_fail(const std::filesystem::path& file_name) {
+  if (std::filesystem::is_directory(file_name)) {
+    return StorageRegion::Mode::FILE_PER_PAGE_SIZE;
+  } else if (std::filesystem::is_block_file(file_name)) {
+    return StorageRegion::Mode::BLOCK;
+  } else {
+    Fail("The backing file has to be either a directory or a block device: " + file_name.string());
+  }
+}
+
+StorageRegion::StorageRegion(const std::filesystem::path& path)
+    : _mode(find_mode_or_fail(path)),
+      _file_handles(_mode == Mode::FILE_PER_PAGE_SIZE ? _open_file_handles_in_directory(path)
+                                                      : _open_file_handles_block(path))
+
+{}
+
+StorageRegion::~StorageRegion() {
+  for (const auto& file_handle : _file_handles) {
+    // We do not handle errors of the close syscall, since we are already in the destructor.
+    close(file_handle.fd);
+    if (_mode == Mode::FILE_PER_PAGE_SIZE) {
+      std::filesystem::remove(file_handle.backing_file_name);
+    }
+  }
+}
+
+StorageRegion::Mode StorageRegion::mode() const {
+  return _mode;
+}
+
+int StorageRegion::_open_file_descriptor(const std::filesystem::path& file_name) {
+  // Partially taken from
+  // DuckDB: https://github.com/duckdb/duckdb/blob/60ed227816669be497fa4ba53e593d3899479c43/src/common/local_file_system.cpp
+  // LevelDB: https://github.com/google/leveldb/commit/296de8d5b8e4e57bd1e46c981114dfbe58a8c4fa
+
+  // On Linux, we use O_DIRECT to bypass the page cache. This is not supported on Mac, but we use the fcntl(F_NOCACHE) call below.
+  // Other nan that, files are opened in read-write mode and created if they do not exist. O_DSYNC ensures that all file data and metadata is
+  // written to the disk before the syscall returns. This eliminates the need for explict fsync calls after each write.
+#ifdef __APPLE__
+  const int flags = O_RDWR | O_CREAT | O_DSYNC;
+#elif __linux__
+  const int flags = O_RDWR | O_CREAT | O_DIRECT | O_DSYNC;
+#endif
+  const int fd = open(file_name.string().c_str(), flags, 0666);
+  if (fd < 0) {
+    const auto error = errno;
+    close(fd);
+    Fail("Error while opening backing file at " + file_name.string().c_str() + ": " + strerror(error) +
+         ". Did you open a file on tmpfs or a network mount?");
```

Comment:
Can you add a comment why tmpfs or network mounts might be problematic? Is it `O_DIRECT`?